### DISCUSSION TITLE
Why SwapChain1 does not have Present1 method?

### DISCUSSION THREAD

**Discussion by MarcioAB** [amerkoleci/Vortice.Windows#438] [Q&A] [answered]
From Microsoft documentation, I understand the method **Present** (p1, p2) is for SwapChain and **Present1** (p1, p2, p3) is for the next interfaces ( SwapChain1, 2, 3 ...).
It seems Vortice.DXGI is using only **Present** for all interfaces.
Is there a reason for that?

**Comment by amerkoleci** [comment]
If you can Present from IDXGISwapChain1 it has more parameters overload, take look here:

https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.DXGI/IDXGISwapChain1.cs#L56

It calls Present1

**Comment by MarcioAB** [comment]
Yes, but even if in the end SwapChain1.Present calls SwapChain1.Present1, SwapChain1.Present1 is **internal** and can not be called directly. It should be public. 
From Microsoft documentation [remarks](https://learn.microsoft.com/en-us/windows/win32/api/dxgi/nf-dxgi-idxgiswapchain-present#remarks) _"Starting with Direct3D 11.1, consider using [IDXGISwapChain1::Present1](https://learn.microsoft.com/en-us/windows/win32/api/dxgi1_2/nf-dxgi1_2-idxgiswapchain1-present1) because you can ..."_

A better title for this Q&A should be: Why SwapChain1 does not have a **public** Present1 method?

**Reply by amerkoleci** [reply]
Not sure I understand what SwapChain1.Present which calls Present1 is missing?

**Comment by MarcioAB** [comment]
Consider someone converting a DX code from C++ to C# with Vortice.
At some point, the C++ code will have a line that calls **Present1**.
This person will try to use the same method ( **Present1** ) but will not find it.
Instead, he will find the "old" **Present** method.
This person will check **Present** in Microsoft documentation and will get the suggestion to use **Present1** instead.
It will be necessary for this person to check the Vortice code to understand **Present** calls an internal **Present1**.
My point is that it will be more straightforward if SwapChain1 has a public **Present1** method and leaves the "old" **Present** method in the "old" SwapChain.

**Reply by amerkoleci** [reply]
Here you go:
https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.DXGI/IDXGISwapChain1.cs#L75

Sounds ok?

**Reply by MarcioAB** [reply]
Yes, but help me to understand:
SwapChain already has a **Present** method.
SwapChain1 is derived from SwapChain, so the **Present** method is already there.
SwapChain1 introduced the new **Present1** method.
I wonder why you created a new **Present** method in SwapChain1 instead of just letting the old one stay in the old SwapChain.
My point is just to keep consistency with DX.
For example: If someone using SwapChain1 wants to use the old Present method, it will not be possible in Vortice, despite it is possible in plain C++ DX.

**Reply by amerkoleci** [reply]
You're sure? You can call both base Present and Present which calls Present1 overload

**Reply by amerkoleci** [reply, marked as answer]
Method renamed to Present1 as suggested, thanks!

https://github.com/amerkoleci/Vortice.Windows/commit/16a632732411b1f3207457738df5463b1cc8add0

Version 3.5.3-beta is comming

**Reply by MarcioAB** [reply]
Great. Thanks. Looking forward.

**Comment by MarcioAB** [comment]
I see: To call Present1 I need to fill the 3 arguments.
```
SWAP12.Present(1) ---> calls Present 
SWAP12.Present(1, PresentFlags.None, new PresentParameters()) ---> calls Present1
```
I can not call Present1 if I just use 1 argument.
That implies Present(n) and Present1(n) are the same.
What if Present1(n) is different from Present(n), for example, faster.

In my code I was using SWAP12.Present(1) expecting I was using SWAP12.Present1(1), but I see now that I was using Present.
As I want to use Present1, even if I do not need those 2 arguments, I will change my code to SWAP12.Present(1, PresentFlags.None, new PresentParameters()) so I'm sure I will be using Present1.

**Reply by amerkoleci** [reply]
If you call Present1 you need to fill PresentParameters otherwise it makes 0 sense, if you don't fill PresentParameters then simply call Present method, you use Present1 to use dirty rectangles and the scroll rectangle in the swap chain presentation and as such use less memory bandwidth and as a result less system power (as said from Microsoft docs)

**Reply by MarcioAB** [reply]
I see. Thank you.
